### PR TITLE
Small-to-Large Merging Solution and Editorial for IOI Race

### PR DESCRIPTION
**Pull request by timothygao8710**
_If any of the below doesn't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which includes the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making them.

### CODE REVIEW

File: solutions/ioi-11-race.mdx [hunk 1]
```diff
@@ -0,0 +1,94 @@
+---
+id: ioi-11-race
+source: IOI 2011
+title: Race
+author: Timothy Gao
+---
+
+[Official Analysis with Centroid Decomposition Downloaded Here](https://ioinformatics.org/page/ioi-2011/37)
```
Comment (SansPapyrus683) [inline solutions/ioi-11-race.mdx:8]:
```suggestion
[Official Analysis (Centroid Decomposition)](https://ioinformatics.org/page/ioi-2011/37)
```

File: solutions/ioi-11-race.mdx [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+#define pb push_back
+ 
+#define f first
+#define s second
+ 
+const int maxn = 2e5+20;
+vector<pii> adj[maxn];
+map<ll, ll> info[maxn]; 
+ll dist[maxn], sum[maxn]; //define these relative to root
+int N, K, u, v; ll ret;
```
Comment (SansPapyrus683) [inline solutions/ioi-11-race.mdx:49]: declarations on separate lines

File: solutions/ioi-11-race.mdx [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+	sum[u] = c; dist[u] = h;
+	for(auto n : adj[u]){
+		if(p == n.f) continue;
+		dfs1(n.f, u, c + n.s, h + 1);
+	}
+}
+ 
+void dfs(int u, int p){
+	ll target = K + 2*sum[u];
+	for(auto n : adj[u]){
```
Comment (SansPapyrus683) [inline solutions/ioi-11-race.mdx:62]: spaces for control flow statements
```suggestion
	for (auto n : adj[u]) {
```
and others

File: solutions/ioi-11-race.mdx [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+ 
+const int maxn = 2e5+20;
+vector<pii> adj[maxn];
+map<ll, ll> info[maxn]; 
+ll dist[maxn], sum[maxn]; //define these relative to root
+int N, K, u, v; ll ret;
+ 
+void dfs1(int u, int p, ll c, int h){
+	info[u][c] = h; 
+	sum[u] = c; dist[u] = h;
```
Comment (SansPapyrus683) [inline solutions/ioi-11-race.mdx:53]: statements on separate lines

File: solutions/ioi-11-race.mdx [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+	for(auto n : adj[u]){
+		if(p == n.f) continue;
```
Comment (SansPapyrus683) [inline solutions/ioi-11-race.mdx:55]: braces (other places too)

File: solutions/ioi-11-race.mdx [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+		dfs1(n.f, u, c + n.s, h + 1);
+	}
+}
+ 
+void dfs(int u, int p){
+	ll target = K + 2*sum[u];
```
Comment (SansPapyrus683) [inline solutions/ioi-11-race.mdx:61]: spaces around operators

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/ioi-11-race.mdx [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+				info[u].insert(i);
+			}else{
+				info[u][i.f] = min(info[u][i.f], i.s);
+			}
+		}
+	} 
+}
+
+int best_path(int n, int k, int edges[][2], int weights[]){
+	if(k == 1) return 0;
```
Comment (ryanchou-dev) [inline solutions/ioi-11-race.mdx:84]: add braces

File: solutions/ioi-11-race.mdx [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+To accomplish this, we can use small to large merging. For each $i$, we maintain a map of key to value pair of $sum(root, j) : dist(root, j)$ for  distinct $sum(root, j)$ for all $j$ in the subtree of $i$. If there are duplicate $sum(root, j)$ we take the minimum $dist(root, j)$.
+
+As we combine the maps of two nodes, we iterate through each key-pair $(sum_s, dist_s)$ of the smaller size map, find if there exists a key-pair $sum_l = s - sum_s$ in the larger map, and obtain $dist_s + dist_l$. Our answer would be minimum $dist_s + dist_l$ over all mergings. During this process we also merge all key-value pairs of the smaller map.
+
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N\log^2N)$
+
+One log factor comes from small-to-large merging, and the other comes from the map operations.
```
Comment (ryanchou-dev) [inline solutions/ioi-11-race.mdx:30]:
```suggestion
The log factors come from small-to-large merging and the map operations.
```

**Review by ryanchou-dev — CHANGES_REQUESTED** [review]
omg hi timothy

File: solutions/ioi-11-race.mdx [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+ll sum[maxn]; // define these relative to root
+int N;
+int K;
+int u;
+int v;
+ll ret;
+
+void dfs1(int u, int p, ll c, int h)
+{
+    info[u][c] = h;
```
Comment (SansPapyrus683) [inline solutions/ioi-11-race.mdx:58]: how could i forget
tabs

File: solutions/ioi-11-race.mdx [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+#define pb push_back
+
+#define f first
+#define s second
+
+const int maxn = 2e5 + 20;
+vector<pii> adj[maxn];
+map<ll, ll> info[maxn];
+ll dist[maxn];
+ll sum[maxn]; // define these relative to root
```
Comment (SansPapyrus683) [inline solutions/ioi-11-race.mdx:49]: two spaces after code before starting an inline comment
```suggestion
ll sum[maxn];  // define these relative to root
```

File: solutions/ioi-11-race.mdx [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+    {
+        if (p == n.f)
+        {
+            continue;
+        }
+        dfs1(n.f, u, c + n.s, h + 1);
+    }
+}
+
+void dfs(int u, int p)
```
Comment (SansPapyrus683) [inline solutions/ioi-11-race.mdx:71]: preferably a better name, what does this `dfs` do?

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/ioi-11-race.mdx [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+
+typedef long long ll;
+typedef pair<ll, ll> pii;
+
+#define pb push_back
+
+#define f first
+#define s second
+
+const int maxn = 2e5 + 20;
```
Comment (SansPapyrus683) [inline solutions/ioi-11-race.mdx:45]: 1. MAX_N instead of maxn
2. would just `2e5` work?

**Review by SansPapyrus683 — APPROVED** [review]
good enough

File: solutions/ioi-11-race.mdx [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+author: Timothy Gao
+---
+
+[Official Analysis (Centroid Decomposition)](https://ioinformatics.org/page/ioi-2011/37)
+
+## Alternative Solution: Small to Large Merging
+
+We are tasked with finding a path with the minimum number of traversed edges such that the path sum is $s$.
+First, we root the tree arbitrarily.
+Let $lca(u, v) = w$.
```
Comment (dongliu0426) [inline solutions/ioi-11-race.mdx:14]: `\texttt lca(u, v)`

File: solutions/ioi-11-race.mdx [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+We can break $path(u, v)$ into $path(u, w)$ and $path(v, w)$.
```
Comment (dongliu0426) [inline solutions/ioi-11-race.mdx:15]: same here, use `\texttt`

Comment (dongliu0426) [reply solutions/ioi-11-race.mdx:15]: not resolved

Comment (timothygao8710) [reply solutions/ioi-11-race.mdx]: added a bunch of \text {}

File: solutions/ioi-11-race.mdx [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+
+const int maxn = 2e5 + 1;
+vector<pii> adj[maxn];
+map<ll, ll> info[maxn];
+ll dist[maxn];
+ll sum[maxn];  // define these relative to root
+int N;
+int K;
+int u;
+int v;
```
[the remark this answers is not in the document]
Comment (SansPapyrus683) [reply solutions/ioi-11-race.mdx]: bruh no

File: solutions/ioi-11-race.mdx [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+
+#define f first
+#define s second
+
+const int maxn = 2e5 + 1;
+vector<pii> adj[maxn];
+map<ll, ll> info[maxn];
+ll dist[maxn];
+ll sum[maxn];  // define these relative to root
+int N, K, u, v;
```
Comment (SansPapyrus683) [inline solutions/ioi-11-race.mdx:50]: wait i don't think `u` and `v` even have to be global?

Comment (timothygao8710) [reply solutions/ioi-11-race.mdx]: true, changed

**Pull request merged**